Explain how to get the discriminant out of a `#[repr(T)] enum` with payload

lukas-code opened the pull request:
example stolen from https://github.com/rust-lang/reference/pull/1055

@rustbot label A-docs

Commit:
explain how to get the discriminant out of a `#[repr(T)] enum`

scottmcm commented on library/core/src/mem/mod.rs:
```diff
@@ -1129,6 +1129,33 @@ impl<T> fmt::Debug for Discriminant<T> {
+///
+/// impl Enum {
+///     fn discriminant(&self) -> u8 {
+///         unsafe { *(self as *const Self as *const u8) }
```
We're trying to avoid `as` on pointers now, where possible, in favour of methods.  So please change this to something using `cast` & friends, maybe
```suggestion
///         unsafe { <*const _>::from(self).cast::<u8>().read() }
```

Or you could consider something like
```rust
    pub fn discriminant(&self) -> &u8 {
        unsafe { NonNull::from(self).cast::<u8>().as_ref() }
    }
```

scottmcm replied:
Oh, and it would be good to have a comment on the `unsafe` block repeating why it's sound.  Maybe

```
// SAFETY: because `Self` is marked `repr(u8)`,
// its layout is a `repr(C)` `union` between `repr(C)` structs,
// each of which has the `u8` discriminant as its first field,
// so we can read the discriminant without offsetting the pointer.
```

lukas-code replied:
TIL that `<*const _>::from(reference)` works. Since there is no `impl From<&T> for *const T`, it probably selects `From<*const T> for *const T` and then coerces the parameter?

We should probably document the "canonical" way to convert a reference to a raw pointer when coercion isn't an option somewhere. So far I've seen `reference as *const _`, `ptr::addr_of!(*reference)` and now `<*const _>::from(reference)`.

lukas-code replied:
I've removed the `as` cast on pointers now, but kept `*ptr` over `ptr.read()` since `u8` is `Copy` and that seems "safer" to me.

scottmcm commented on library/core/src/mem/mod.rs:
````diff
@@ -1129,6 +1129,33 @@ impl<T> fmt::Debug for Discriminant<T> {
 /// ```
+///
+/// Note that it is *undefined behavior* to [`transmute`] from [`Discriminant`] to a primitive.
+/// To access the value of a discriminant with a primitive representation, use pointer casts instead:
````
Note that the test below currently "passes" [even without the `repr`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=994ac84af453c7547ad6e0854742bb8e), though it's UB without the `repr` -- and MIRI doesn't complain about it either.

So thus I think it's important to put extra emphasis on how this only applies for non-default `repr`s, and it'd be nice to link to some documentation about those other reprs.

Here's a first draft of an idea, please improve it:

> Note that it is *undefined behavior* to [`transmute`] from [`Discriminant`] to a primitive.
>
> If an enum has opted-in to having a [defined layout](https://rust-lang.github.io/rfcs/2195-really-tagged-unions.html#guide-level-explanation), then it's possible to use pointers to read the memory location storing the discriminant.  That **cannot** be done for anything using the *default* `enum` layout, however, as it's undefined where the discriminant is stored -- it might not even be stored at all!

scottmcm reviewed (changes requested):
Thanks for the PR!  Overall it looks good, though I have a few hopefully-simple requests.

@rustbot author

lukas-code commented on library/core/src/mem/mod.rs:
```diff
@@ -1113,7 +1113,10 @@ impl<T> fmt::Debug for Discriminant<T> {
+/// of some variant will not change between compilations with the same compiler. See the [Reference]
+/// for more information.
+///
+/// [Reference]: ../../reference/items/enumerations.html#discriminants
```
This section doesn't exist until https://github.com/rust-lang/reference/pull/1055 is merged, but [the linked page](https://doc.rust-lang.org/nightly/reference/items/enumerations.html#discriminants) still contains some relevant information.

lukas-code replied:
The CI didn't like this, so we need to revert 946d51e8ba2501aa24256317d8798e475dc862b5 later.

lukas-code commented:
I've updated it now to include some links to the Reference on enum reprs. I'd rather not link to the RFC, since RFCs tend to get outdated over time and that [has led to confusion before](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-lang.2Fwg-unsafe-code-guidelines/topic/RFC.202195.20and.20enum.20validity). I intentionally left out `repr(C)` enums here, since the ABI of C enums is notably underspecified and may or may not be `c_int`.

I also added the easy way to get the discriminant of a c-like enum now, to avoid having someone use unsafe code when it isn't necessary.

Docs preview available here: https://lukas-code.github.io/rust-docs/core/mem/fn.discriminant.html

lukas-code commented:
@rustbot ready

Commit:
fix broken link fragment

scottmcm commented:
Thanks for the updates!  Having the safe way there first is a great addition.

@bors r+ rollup

bors commented:
:pushpin: Commit 946d51e8ba2501aa24256317d8798e475dc862b5 has been approved by `scottmcm`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=scottmcm 946d51e8ba2501aa24256317d8798e475dc862b5 -->
<!-- homu: {"type":"Approved","sha":"946d51e8ba2501aa24256317d8798e475dc862b5","approver":"scottmcm","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
